core/src/main/java/com/linecorp/armeria/server/RouteDecoratingService.java

```diff
@@ -0,0 +1,111 @@
+
+    private static final AttributeKey<Queue<Service<HttpRequest, HttpResponse>>> DECORATOR_KEY =
+            AttributeKey.valueOf(RouteDecoratingService.class, "PENDING_DECORATORS");
+
+    static Function<Service<HttpRequest, HttpResponse>,
+            Service<HttpRequest, HttpResponse>> newDecorator(Router<RouteDecoratingService> router) {
+        return delegate -> new InitialDispatcherService(delegate, router);
+    }
+
+    private final Route route;
+    private final Service<HttpRequest, HttpResponse> decorator;
+
+    RouteDecoratingService(Route route,
+                           Function<Service<HttpRequest, HttpResponse>,
+                                   ? extends Service<HttpRequest, HttpResponse>> decoratorFunction) {
+        this.route = requireNonNull(route, "route");
+        decorator = requireNonNull(decoratorFunction, "decoratorFunction").apply(this);
+    }
+
+    @Override
+    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        final Queue<Service<HttpRequest, HttpResponse>> delegates = ctx.attr(DECORATOR_KEY).get();
+        final Service<HttpRequest, HttpResponse> delegate = delegates.poll();
+        return delegate.serve(ctx, req);
+    }
+
+
+    Route route() {
+        return route;
+    }
+
+    private Service<HttpRequest, HttpResponse> decorator() {
+        return decorator;
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                          .add("route", route)
+                          .add("decorator", decorator).toString();
+    }
+
+    private static class InitialDispatcherService extends SimpleDecoratingHttpService {
+
+        private final Router<RouteDecoratingService> router;
+
+        InitialDispatcherService(Service<HttpRequest, HttpResponse> delegate,
+                                 Router<RouteDecoratingService> router) {
+            super(delegate);
+            this.router = router;
+        }
+
+        @Override
+        public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+            final Queue<Service<HttpRequest, HttpResponse>> pendingDecorators = new ArrayDeque<>();
+            for (Routed<RouteDecoratingService> routed : router.findAll(ctx.routingContext())) {
+                if (routed.isPresent()) {
+                    pendingDecorators.add(routed.value().decorator());
+                }
+            }
+
+            if (pendingDecorators.isEmpty()) {
+                return delegate().serve(ctx, req);
+            }
+            pendingDecorators.add(delegate());
+            final Service<HttpRequest, HttpResponse> decorator = pendingDecorators.poll();
+            ctx.attr(DECORATOR_KEY).set(pendingDecorators);
+            return decorator.serve(ctx, req);
+        }
+
+        @Override
+        public String toString() {
+            return "InitialDispatcherService()";
```

Comment:
I guess this will make most route-decorated services will just return `InitialDispatcherService()` if we do this. How about this:

    return "(..., " + delegate() + ')';

Not sure it's the best, so please feel free to do as you want. I'm fine as long as `delegate()` is part of the output.

Reply:
I used `MoreObjects.toStringHelper()`.